### PR TITLE
XXE Level - Integration Tests - Dev#334

### LINKED ISSUE #334: Investigate creating XXE challenges

### LINKED ISSUE #496: [BUG] Setup Page not displaying MongoDb setup properly
**Describe the bug**
i18n not displaying properly on the setup page

**To Reproduce**
1. Load the setup page
2. observer the following
???generic.text.setup.enable??? MongoDb Challenge 

**Expected behavior**
Should display text properly 

**Screenshots**
image

### PR DESCRIPTION
**Pull request by ismisepaul**
- Added XXE Level - closes #334
- Added ability to mark levels as "unsafe". Any future levels that make the server vulnerable can be marked as such. Unsafe levels can be enabled or disabled at setup or by the admin in the configuration. 
- Fixed MongoDb issues on setup page - closes #496
- Separated unit tests from integration tests so they can be run separately in travis-ci

### CODE REVIEW

File: src/it/java/servlets/admin/moduleManagement/OpenAllModulesIT.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+            assertTrue(responseBody.contains("[WARNING] Server is vulnerable. Unsafe levels open!"));
+            assertTrue(responseBody.contains("All Modules are Now Open"));
+
+        }
+        catch (NullPointerException e) { fail("Null Pointer" + e.toString());}
+        catch (Exception e) { fail(e.toString()); }
+    }
+
+}
```
Comment (markdenihan) [inline src/it/java/servlets/admin/moduleManagement/OpenAllModulesIT.java:179]: Perhaps an Unauthenticated test as well considering the danger of the function?

File: src/it/java/servlets/module/lesson/XxeLessonIT.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		catch(Exception e)
+		{
+			log.fatal("Could not Complete: " + e.toString());
+			fail("Could not Complete: " + e.toString());
+		}
+	}
+
+
+
+}
```
Comment (markdenihan) [inline src/it/java/servlets/module/lesson/XxeLessonIT.java:179]: Maybe add a test here to see if the level doesn't work if Dangerous Levels have been disabled?

File: src/main/java/dbProcs/Setter.java [hunk 3]
```diff
@@ -127,17 +148,17 @@ public static boolean incrementBadSubmission(String ApplicationRoot, String user
 	/**
 	 * This method sets every module status to Open.
 	 * @param ApplicationRoot Current running director of the application
-	 * @param moduleId The identifier of the module that is been set to open status
+	 * @param unsafe set whether to open all modules or all modules that are safe
 	 * @return Boolean result depicting success of statement
 	 */
-	public static boolean openAllModules (String ApplicationRoot)
+	public static boolean openAllModules (String ApplicationRoot, int unsafe)
```
Comment (markdenihan) [inline src/main/java/dbProcs/Setter.java]: Maybe a boolean instead of an int?

**Review by markdenihan — CHANGES_REQUESTED** [review]

**Comment by jmanico** [comment]
XXE?? What version of Shepherd includes this???

--
Jim Manico
@Manicode

**Comment by ismisepaul** [comment]
> XXE?? What version of Shepherd includes this???
> […](#)
> -- Jim Manico @manicode


@jmanico we haven't released it yet. Still working on the XXE branch. As soon as I tidy up the merge we'll try get it out in the next version of Shepherd.

**Comment by jmanico** [comment]
Awesome!

--
Jim Manico
@Manicode
Secure Coding Education
+1 (808) 652-3805

File: src/it/java/servlets/module/lesson/XxeLessonIT.java (hunk 2 above)
Comment (ismisepaul) [reply src/it/java/servlets/module/lesson/XxeLessonIT.java:179]: Test has been added. Also code added to the level to not allow user to submit anything if that level is closed.

File: src/it/java/servlets/admin/moduleManagement/OpenAllModulesIT.java (hunk 1 above)
Comment (ismisepaul) [reply src/it/java/servlets/admin/moduleManagement/OpenAllModulesIT.java:179]: Test has been added

File: src/main/java/dbProcs/Setter.java (hunk 3 above)
Comment (ismisepaul) [reply src/main/java/dbProcs/Setter.java]: boolean used now instead of int

File: src/main/java/servlets/module/lesson/XxeLesson.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+ * along with the Security Shepherd project.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @author ismisepaul
+ */
+public class XxeLesson
+        extends HttpServlet {
+    private static final long serialVersionUID = 1L;
+    private static Logger log = Logger.getLogger(XxeLesson.class);
+    private static final String LEVEL_NAME = "XXE Lesson";
+    private static final String LEVEL_ID = "df2ac757cc135dcb8ce5ea01f677c74f04b446d6";
```
Comment (markdenihan) [inline src/main/java/servlets/module/lesson/XxeLesson.java]: Should this be pulled out of the DB via using the Hash? I don't think Level ID's are hard coded anywhere else off the top of my head

File: src/main/java/servlets/module/lesson/XxeLesson.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+                        String htmlOutput = new String();
+
+                        if (emailAddr == null) {
+                            htmlOutput += "<p>" + bundle.getString("response.blank.email") + "</p>";
+                            out.write(htmlOutput + emailAddr);
+                        } else if (Validate.isValidEmailAddress(emailAddr)) {
+                            log.debug("User Submitted - " + emailAddr);
+
+                            htmlOutput += "<p>" + bundle.getString("response.success.reset") + ": " + emailAddr
```
Comment (markdenihan) [inline src/main/java/servlets/module/lesson/XxeLesson.java:96]: Shouldn't emailAddr be encoded for HTML before added to the htmlOutput string?

File: src/main/java/servlets/module/lesson/XxeLesson.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+                                    + " has been reset</p>";
+                            out.write(htmlOutput);
+                        } else {
+                            htmlOutput += "<p>" + bundle.getString("response.invalid.email") + ": "
+                                    + emailAddr + "</p>";
```
Comment (markdenihan) [inline src/main/java/servlets/module/lesson/XxeLesson.java:101]: Shouldn't emailAddr be encoded for HTML before added to the htmlOutput string?

**Review by markdenihan — CHANGES_REQUESTED** [review]

File: src/main/java/servlets/module/lesson/XxeLesson.java (hunk 5 above)
Comment (ismisepaul) [reply src/main/java/servlets/module/lesson/XxeLesson.java:96]: The encoding is being done in the readXml() function in that class. The return function encodes it.

set 
https://github.com/OWASP/SecurityShepherd/blob/2cac9ef1673aa8716253a645be66ef3db4145a58/src/main/java/servlets/module/lesson/XxeLesson.java#L85

encoded
https://github.com/OWASP/SecurityShepherd/blob/2cac9ef1673aa8716253a645be66ef3db4145a58/src/main/java/servlets/module/lesson/XxeLesson.java#L135

File: src/main/java/servlets/module/lesson/XxeLesson.java (hunk 6 above)
Comment (ismisepaul) [reply src/main/java/servlets/module/lesson/XxeLesson.java:101]: The encoding is being done in the readXml() function in that class. The return function encodes it.

set 
https://github.com/OWASP/SecurityShepherd/blob/2cac9ef1673aa8716253a645be66ef3db4145a58/src/main/java/servlets/module/lesson/XxeLesson.java#L85

encoded
https://github.com/OWASP/SecurityShepherd/blob/2cac9ef1673aa8716253a645be66ef3db4145a58/src/main/java/servlets/module/lesson/XxeLesson.java#L135

File: src/main/java/servlets/module/lesson/XxeLesson.java (hunk 4 above)
Comment (ismisepaul) [reply src/main/java/servlets/module/lesson/XxeLesson.java]: Switched to grabbing it from the level hash (yeah it's not being done anywhere else)

https://github.com/OWASP/SecurityShepherd/blob/2cac9ef1673aa8716253a645be66ef3db4145a58/src/main/java/servlets/module/lesson/XxeLesson.java#L58

**Review by markdenihan — APPROVED** [review]

**Pull request merged**